tools/jnimarshalmethod-gen/App.cs

```diff
@@ -56,43 +114,96 @@ static void CreateMarshalMethodAssembly (string path)
 			var destPath        = assemblyName.Name + ".dll";
 			var builder         = CreateExportedMemberBuilder ();
 
+			if (verbose)
+				Console.WriteLine ($"Preparing marshal method assembly '{assemblyName}'");
+
 			var da = AppDomain.CurrentDomain.DefineDynamicAssembly (
 					assemblyName,
 					AssemblyBuilderAccess.Save,
 					Path.GetDirectoryName (path));
 
 			var dm = da.DefineDynamicModule ("<default>", destPath);
 
+			var ad = resolver.GetAssembly (path);
+
 			foreach (var type in assembly.DefinedTypes) {
+				if (type.IsGenericType || type.IsGenericTypeDefinition)
+					continue;
+
+				var td                      = ad.MainModule.FindType (type);
+
+				if (td == null) {
+					if (verbose)
+						Warning ($"Unable to find cecil's TypeDefinition of type {type}");
+					continue;
+				}
+
 				var registrationElements    = new List<Expression> ();
 				var targetType              = Expression.Variable (typeof(Type), "targetType");
 				TypeBuilder dt = null;
 
 				var flags = BindingFlags.Public | BindingFlags.NonPublic |
 						BindingFlags.Instance | BindingFlags.Static;
-				foreach (var method in type.GetMethods (flags )) {
-					// TODO: Constructors, [Register] methods
+				foreach (var method in type.GetMethods (flags)) {
+					// TODO: Constructors
 					var export  = method.GetCustomAttribute<JavaCallableAttribute> ();
-					if (export == null)
-						continue;
+					string signature = null;
+					string name = null;
+					string methodName = method.Name;
+
+					if (export == null) {
+						if (method.IsGenericMethod || method.ContainsGenericParameters || method.IsGenericMethodDefinition || method.ReturnType.IsGenericType)
+							continue;
+
+						if (method.DeclaringType != type || !td.ImplementsInterface ("Java.Interop.IJavaPeerable"))
```

Comment:
The `!td.ImplementsInterface()` should be done earlier. If the type doesn't implement `Java.Interop.IJavaPeerable`, then we don't need to do *anything*. We could thus bail earlier, before we check *any* methods.